v2.0.0-alpha-2: Support for ES3/5 'modules' (implicit strict mode) is lost

`ecmaFeatures.modules = true` in v1.10.3 supports ES3/5 'modules' as well as ES6 modules, whereas `parserOptions.sourceType = "module"` in v2.0.0-alpha-2 only supports ES6 modules (because it silently overrides `parserOptions.ecmaVersion`: https://github.com/eslint/eslint/blob/master/lib/eslint.js#L418).

An ES3/5 'module' may be *implicitly* in strict mode. For example, this file:
```js
// myModule.js
exports.foo = "bar";
```
is implicitly in strict mode because it will be bundled up into something like this:
```js
// bundle.js
loadModules({
    myModule: function (require, exports, module) {
        "use strict";
        // myModule.js
        exports.foo = "bar";
    },
});
```

Support for ES3/5 'modules' and implicit strict mode can be tested with the following files, both of which contain an error:
```js
// es6.js
const ES6 = true;
```
```js
// invalidThis.js
function invalidThis() {
    return this;
}
```

v1.10.3 supports ES3/5 'modules' because it finds both errors:
```json
{
    "ecmaFeatures": {
        "modules": true
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
```
es6.js
  2:2  error  Parsing error: Unexpected token const

invalidThis.js
  3:9  error  Unexpected `this`  no-invalid-this

✖ 2 problems (2 errors, 0 warnings)
```

v2.0.0-alpha-2 does not support ES3/5 'modules' because it can find either the first error:
```json
{
    "parserOptions": {
        "ecmaVersion": 5
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
```
es6.js
  2:2  error  Parsing error: The keyword 'const' is reserved

✖ 1 problem (1 error, 0 warnings)
```
or the second error (but not both):
```json
{
    "parserOptions": {
        "ecmaVersion": 5,
        "sourceType": "module"
    },
    "rules": {
        "no-invalid-this": 2
    }
}
```
```
invalidThis.js
  3:9  error  Unexpected `this`  no-invalid-this

✖ 1 problem (1 error, 0 warnings)
```

Although it is correct that `parserOptions.sourceType = "module"` should imply ES6 (although perhaps not silently), I hope that support for ES3/5 'modules' and implicit strict mode can be restored to v2 in some way.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29482221-v2-0-0-alpha-2-support-for-es3-5-modules-implicit-strict-mode-is-lost?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/docs/rules/strict.md
+++ b/docs/rules/strict.md
@@ -35,7 +35,7 @@ There are four options for this rule:
 1. `"function"` - require `"use strict"` in function scopes only
 1. `"safe"` - require `"use strict"` globally when inside a module wrapper and in function scopes everywhere else.
 
-All strict mode directives are flagged as unnecessary if ECMAScript modules are enabled (see [Specifying Parser Options](../user-guide/configuring#specifying-parser-options)). This behaviour does not depend on the rule options, but can be silenced by disabling this rule.
+All strict mode directives are flagged as unnecessary if ECMAScript modules or implied strict mode are enabled (see [Specifying Parser Options](../user-guide/configuring#specifying-parser-options)). This behaviour does not depend on the rule options, but can be silenced by disabling this rule.
 
 ### "never"
 

--- a/docs/user-guide/configuring.md
+++ b/docs/user-guide/configuring.md
@@ -23,6 +23,7 @@ Parser options are set in your `.eslintrc.*` file by using the `parserOptions` p
 * `sourceType` - set to `"script"` (default) or `"module"` if your code is in ECMAScript modules.
 * `ecmaFeatures` - an object indicating which additional language features you'd like to use:
     * `globalReturn` - allow `return` statements in the global scope
+    * `impliedStrict` - enable global [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) (if `ecmaVersion` is 5 or greater)
     * `jsx` - enable [JSX](http://facebook.github.io/jsx/)
     * `experimentalObjectRestSpread` - enable support for the experimental [object rest/spread properties](https://github.com/sebmarkbage/ecmascript-rest-spread) (**IMPORTANT:** This is an experimental feature that may change significantly in the future. It's recommended that you do *not* write rules relying on this functionality unless you are willing to incur maintenance cost when it changes.)
 

--- a/lib/eslint.js
+++ b/lib/eslint.js
@@ -777,6 +777,7 @@ module.exports = (function() {
             scopeManager = escope.analyze(ast, {
                 ignoreEval: true,
                 nodejsScope: ecmaFeatures.globalReturn,
+                impliedStrict: ecmaFeatures.impliedStrict,
                 ecmaVersion: ecmaVersion,
                 sourceType: currentConfig.parserOptions.sourceType || "script"
             });

--- a/lib/rules/strict.js
+++ b/lib/rules/strict.js
@@ -25,6 +25,7 @@ var messages = {
     never: "Strict mode is not permitted.",
     unnecessary: "Unnecessary 'use strict' directive.",
     module: "'use strict' is unnecessary inside of modules.",
+    implied: "'use strict' is unnecessary when implied strict mode is enabled.",
     unnecessaryInClasses: "'use strict' is unnecessary inside of classes."
 };
 
@@ -62,14 +63,15 @@ function getUseStrictDirectives(statements) {
 module.exports = function(context) {
 
     var mode = context.options[0] || "safe",
+        ecmaFeatures = context.parserOptions.ecmaFeatures || {},
         scopes = [],
         classScopes = [],
         rule;
 
-    if (mode === "safe") {
-        mode = context.parserOptions.ecmaFeatures &&
-            context.parserOptions.ecmaFeatures.globalReturn ?
-            "global" : "function";
+    if (ecmaFeatures.impliedStrict) {
+        mode = "implied";
+    } else if (mode === "safe") {
+        mode = ecmaFeatures.globalReturn ? "global" : "function";
     }
 
     /**

--- a/package.json
+++ b/package.json
@@ -41,7 +41,7 @@
     "doctrine": "^1.1.0",
     "es6-map": "^0.1.3",
     "escape-string-regexp": "^1.0.2",
-    "escope": "^3.3.0",
+    "escope": "^3.4.0",
     "espree": "^3.0.0",
     "estraverse": "^4.1.1",
     "estraverse-fb": "^1.3.1",

--- a/tests/lib/eslint.js
+++ b/tests/lib/eslint.js
@@ -2736,6 +2736,40 @@ describe("eslint", function() {
             assert.equal(messages[0].message, "Parsing error: 'return' outside of function");
         });
 
+        it("should properly parse sloppy-mode code when impliedStrict is false", function() {
+
+            var messages = eslint.verify("var private;", {}, filename);
+
+            assert.equal(messages.length, 0);
+        });
+
+        it("should not parse sloppy-mode code when impliedStrict is true", function() {
+
+            var messages = eslint.verify("var private;", {
+                parserOptions: {
+                    ecmaFeatures: {
+                        impliedStrict: true
+                    }
+                }
+            }, filename);
+
+            assert.equal(messages.length, 1);
+            assert.equal(messages[0].message, "Parsing error: The keyword 'private' is reserved");
+        });
+
+        it("should properly parse valid code when impliedStrict is true", function() {
+
+            var messages = eslint.verify("var foo;", {
+                parserOptions: {
+                    ecmaFeatures: {
+                        impliedStrict: true
+                    }
+                }
+            }, filename);
+
+            assert.equal(messages.length, 0);
+        });
+
         it("should properly parse JSX when passed ecmaFeatures", function() {
 
             var messages = eslint.verify("var x = <div/>;", {

--- a/tests/lib/rules/no-eval.js
+++ b/tests/lib/rules/no-eval.js
@@ -40,6 +40,7 @@ ruleTester.run("no-eval", rule, {
         "this.noeval('foo');",
         "function foo() { 'use strict'; this.eval('foo'); }",
         { code: "function foo() { this.eval('foo'); }", parserOptions: { sourceType: "module" } },
+        { code: "function foo() { this.eval('foo'); }", parserOptions: { ecmaFeatures: { impliedStrict: true } } },
         "var obj = {foo: function() { this.eval('foo'); }}",
         "var obj = {}; obj.foo = function() { this.eval('foo'); }",
         { code: "class A { foo() { this.eval(); } }", parserOptions: { ecmaVersion: 6 } },

--- a/tests/lib/rules/no-invalid-this.js
+++ b/tests/lib/rules/no-invalid-this.js
@@ -10,7 +10,7 @@
 // Requirements
 //------------------------------------------------------------------------------
 
-var assign = require("object-assign");
+var cloneDeep = require("lodash.clonedeep");
 var rule = require("../../../lib/rules/no-invalid-this");
 var RuleTester = require("../../../lib/testers/rule-tester");
 
@@ -37,6 +37,18 @@ function USE_STRICT(pattern) {
 }
 
 /**
+ * A constant value for implied strict mode.
+ * This modifies pattern object to impose strict mode.
+ * @param {object} pattern - A pattern object to modify.
+ * @returns {void}
+ */
+function IMPLIED_STRICT(pattern) {
+    pattern.code = "/* implied strict mode */ " + pattern.code;
+    pattern.parserOptions.ecmaFeatures = pattern.parserOptions.ecmaFeatures || {};
+    pattern.parserOptions.ecmaFeatures.impliedStrict = true;
+}
+
+/**
  * A constant value for modules environment.
  * This modifies pattern object to make modules.
  * @param {object} pattern - A pattern object to modify.
@@ -44,7 +56,7 @@ function USE_STRICT(pattern) {
  */
 function MODULES(pattern) {
     pattern.code = "/* modules */ " + pattern.code;
-    pattern.parserOptions = assign({}, pattern.parserOptions, {sourceType: "module"});
+    pattern.parserOptions.sourceType = "module";
 }
 
 /**
@@ -57,7 +69,7 @@ function extractPatterns(patterns, type) {
     // Clone and apply the pattern environment.
     var patternsList = patterns.map(function(pattern) {
         return pattern[type].map(function(applyCondition) {
-            var thisPattern = assign({}, pattern);
+            var thisPattern = cloneDeep(pattern);
             applyCondition(thisPattern);
 
             if (type === "valid") {
@@ -91,7 +103,7 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "console.log(this); z(x => console.log(x, this));",
@@ -101,7 +113,7 @@ var patterns = [
         },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
 
     // IIFE.
@@ -110,7 +122,7 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
 
     // Just functions.
@@ -119,14 +131,14 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "function foo() { \"use strict\"; console.log(this); z(x => console.log(x, this)); }",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [],
-        invalid: [NORMAL, USE_STRICT, MODULES]
+        invalid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "return function() { console.log(this); z(x => console.log(x, this)); };",
@@ -136,14 +148,14 @@ var patterns = [
         },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT] // modules cannot return on global.
+        invalid: [USE_STRICT, IMPLIED_STRICT] // modules cannot return on global.
     },
     {
         code: "var foo = (function() { console.log(this); z(x => console.log(x, this)); }).bar(obj);",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
 
     // Functions in methods.
@@ -152,49 +164,49 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "var obj = {foo() { function foo() { console.log(this); z(x => console.log(x, this)); } foo(); }};",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "var obj = {foo: function() { return function() { console.log(this); z(x => console.log(x, this)); }; }};",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "var obj = {foo: function() { \"use strict\"; return function() { console.log(this); z(x => console.log(x, this)); }; }};",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [],
-        invalid: [NORMAL, USE_STRICT, MODULES]
+        invalid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "obj.foo = function() { return function() { console.log(this); z(x => console.log(x, this)); }; };",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "obj.foo = function() { \"use strict\"; return function() { console.log(this); z(x => console.log(x, this)); }; };",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [],
-        invalid: [NORMAL, USE_STRICT, MODULES]
+        invalid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "class A { foo() { return function() { console.log(this); z(x => console.log(x, this)); }; } }",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [],
-        invalid: [NORMAL, USE_STRICT, MODULES]
+        invalid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES]
     },
 
     // Class Static methods.
@@ -202,7 +214,7 @@ var patterns = [
         code: "class A {static foo() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -210,19 +222,19 @@ var patterns = [
     {
         code: "function Foo() { console.log(this); z(x => console.log(x, this)); }",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "var Foo = function Foo() { console.log(this); z(x => console.log(x, this)); };",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "class A {constructor() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -230,43 +242,43 @@ var patterns = [
     {
         code: "var obj = {foo: function() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "var obj = {foo() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "var obj = {foo: foo || function() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "var obj = {foo: hasNative ? foo : function() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "var obj = {foo: (function() { return function() { console.log(this); z(x => console.log(x, this)); }; })()};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "Object.defineProperty(obj, \"foo\", {value: function() { console.log(this); z(x => console.log(x, this)); }})",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "Object.defineProperties(obj, {foo: {value: function() { console.log(this); z(x => console.log(x, this)); }}})",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -274,25 +286,25 @@ var patterns = [
     {
         code: "obj.foo = function() { console.log(this); z(x => console.log(x, this)); };",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "obj.foo = foo || function() { console.log(this); z(x => console.log(x, this)); };",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "obj.foo = foo ? bar : function() { console.log(this); z(x => console.log(x, this)); };",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "obj.foo = (function() { return function() { console.log(this); z(x => console.log(x, this)); }; })();",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -300,7 +312,7 @@ var patterns = [
     {
         code: "class A {foo() { console.log(this); z(x => console.log(x, this)); }};",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -308,7 +320,7 @@ var patterns = [
     {
         code: "var foo = function() { console.log(this); z(x => console.log(x, this)); }.bind(obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
@@ -316,12 +328,12 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "(function() { console.log(this); z(x => console.log(x, this)); }).call(obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
@@ -329,12 +341,12 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "(function() { console.log(this); z(x => console.log(x, this)); }).apply(obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
@@ -342,12 +354,12 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "Reflect.apply(function() { console.log(this); z(x => console.log(x, this)); }, obj, []);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -357,103 +369,103 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.every(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.filter(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.find(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.findIndex(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.forEach(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.map(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo.some(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "Array.from([], function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.every(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.filter(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.find(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.findIndex(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.forEach(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.map(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "foo.some(function() { console.log(this); z(x => console.log(x, this)); }, obj);",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
@@ -461,20 +473,20 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
 
     // @this tag.
     {
         code: "/** @this Obj */ function foo() { console.log(this); z(x => console.log(x, this)); }",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
         code: "/**\n * @returns {void}\n * @this Obj\n */\nfunction foo() { console.log(this); z(x => console.log(x, this)); }",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
     {
@@ -482,20 +494,20 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "/** @this Obj */ foo(function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
     {
         code: "foo(/* @this Obj */ function() { console.log(this); z(x => console.log(x, this)); });",
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     },
 
@@ -505,14 +517,14 @@ var patterns = [
         parserOptions: { ecmaVersion: 6 },
         errors: errors,
         valid: [NORMAL],
-        invalid: [USE_STRICT, MODULES]
+        invalid: [USE_STRICT, IMPLIED_STRICT, MODULES]
     },
 
     // https://github.com/eslint/eslint/issues/3287
     {
         code: "function foo() { /** @this Obj*/ return function bar() { console.log(this); z(x => console.log(x, this)); }; }",
         parserOptions: { ecmaVersion: 6 },
-        valid: [NORMAL, USE_STRICT, MODULES],
+        valid: [NORMAL, USE_STRICT, IMPLIED_STRICT, MODULES],
         invalid: []
     }
 ];

--- a/tests/lib/rules/no-lone-blocks.js
+++ b/tests/lib/rules/no-lone-blocks.js
@@ -27,6 +27,7 @@ ruleTester.run("no-lone-blocks", rule, {
         {code: "{ let x = 1; }", parserOptions: { ecmaVersion: 6 }},
         {code: "{ const x = 1; }", parserOptions: { ecmaVersion: 6 }},
         {code: "'use strict'; { function bar() {} }", parserOptions: { ecmaVersion: 6 }},
+        {code: "{ function bar() {} }", parserOptions: { ecmaVersion: 6, ecmaFeatures: { impliedStrict: true }}},
         {code: "{ class Bar {} }", parserOptions: { ecmaVersion: 6 }},
 
         {code: "{ {let y = 1;} let x = 1; }", parserOptions: { ecmaVersion: 6 }}

--- a/tests/lib/rules/strict.js
+++ b/tests/lib/rules/strict.js
@@ -28,11 +28,13 @@ ruleTester.run("strict", rule, {
         { code: "var fn = x => 1;", parserOptions: { ecmaVersion: 6 }, options: ["never"] },
         { code: "var fn = x => { return; };", parserOptions: { ecmaVersion: 6 }, options: ["never"] },
         { code: "foo();", parserOptions: { sourceType: "module" }, options: ["never"] },
+        { code: "function foo() { return; }", parserOptions: { ecmaFeatures: { impliedStrict: true } }, options: ["never"] },
 
         // "global" mode
         { code: "// Intentionally empty", options: ["global"] },
         { code: "\"use strict\"; foo();", options: ["global"] },
         { code: "foo();", parserOptions: { sourceType: "module" }, options: ["global"] },
+        { code: "function foo() { return; }", parserOptions: { ecmaFeatures: { impliedStrict: true } }, options: ["global"] },
         { code: "'use strict'; function foo() { return; }", options: ["global"] },
         { code: "'use strict'; var foo = function() { return; };", options: ["global"] },
         { code: "'use strict'; function foo() { bar(); 'use strict'; return; }", options: ["global"] },
@@ -43,6 +45,7 @@ ruleTester.run("strict", rule, {
         // "function" mode
         { code: "function foo() { 'use strict'; return; }", options: ["function"] },
         { code: "function foo() { return; }", parserOptions: { sourceType: "module" }, options: ["function"] },
+        { code: "function foo() { return; }", parserOptions: { ecmaFeatures: { impliedStrict: true } }, options: ["function"] },
         { code: "var foo = function() { return; }", parserOptions: { sourceType: "module" }, options: ["function"] },
         { code: "var foo = function() { 'use strict'; return; }", options: ["function"] },
         { code: "function foo() { 'use strict'; return; } var bar = function() { 'use strict'; bar(); };", options: ["function"] },
@@ -69,10 +72,14 @@ ruleTester.run("strict", rule, {
         // "safe" mode corresponds to "global" if ecmaFeatures.globalReturn is true, otherwise "function"
         { code: "function foo() { 'use strict'; return; }", options: ["safe"] },
         { code: "'use strict'; function foo() { return; }", parserOptions: { ecmaFeatures: { globalReturn: true } }, options: ["safe"] },
+        { code: "function foo() { return; }", parserOptions: { sourceType: "module" }, options: ["safe"] },
+        { code: "function foo() { return; }", parserOptions: { ecmaFeatures: { impliedStrict: true } }, options: ["safe"] },
 
         // defaults to "safe" mode
         { code: "function foo() { 'use strict'; return; }" },
-        { code: "'use strict'; function foo() { return; }", parserOptions: { ecmaFeatures: { globalReturn: true } } }
+        { code: "'use strict'; function foo() { return; }", parserOptions: { ecmaFeatures: { globalReturn: true } } },
+        { code: "function foo() { return; }", parserOptions: { sourceType: "module" } },
+        { code: "function foo() { return; }", parserOptions: { ecmaFeatures: { impliedStrict: true } } }
 
     ],
     invalid: [
@@ -114,6 +121,22 @@ ruleTester.run("strict", rule, {
             options: ["never"],
             parserOptions: { sourceType: "module" },
             errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
+            ]
+        }, {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["never"],
+            parserOptions: { ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" }
+            ]
+        }, {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["never"],
+            parserOptions: { sourceType: "module", ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" },
                 { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
             ]
         },
@@ -170,6 +193,22 @@ ruleTester.run("strict", rule, {
             options: ["global"],
             parserOptions: { sourceType: "module" },
             errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
+            ]
+        }, {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["global"],
+            parserOptions: { ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" }
+            ]
+        }, {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["global"],
+            parserOptions: { sourceType: "module", ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" },
                 { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
             ]
         },
@@ -237,6 +276,22 @@ ruleTester.run("strict", rule, {
             options: ["function"],
             parserOptions: { sourceType: "module" },
             errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
+            ]
+        }, {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["function"],
+            parserOptions: { ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" }
+            ]
+        }, {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["function"],
+            parserOptions: { sourceType: "module", ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" },
                 { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
             ]
         }, {
@@ -323,6 +378,24 @@ ruleTester.run("strict", rule, {
                 { message: "Use the global form of 'use strict'.", type: "ExpressionStatement" }
             ]
         },
+        {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["safe"],
+            parserOptions: { ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" }
+            ]
+        },
+        {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            options: ["safe"],
+            parserOptions: { sourceType: "module", ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
+            ]
+        },
 
         // Default to "safe" mode
         {
@@ -342,6 +415,22 @@ ruleTester.run("strict", rule, {
             errors: [
                 { message: "Use the global form of 'use strict'.", type: "Program" },
                 { message: "Use the global form of 'use strict'.", type: "ExpressionStatement" }
+            ]
+        },
+        {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            parserOptions: { ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary when implied strict mode is enabled.", type: "ExpressionStatement" }
+            ]
+        },
+        {
+            code: "'use strict'; function foo() { 'use strict'; return; }",
+            parserOptions: { sourceType: "module", ecmaFeatures: { impliedStrict: true } },
+            errors: [
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" },
+                { message: "'use strict' is unnecessary inside of modules.", type: "ExpressionStatement" }
             ]
         }
 

